pennylane/capture/base_interpreter.py

```diff
@@ -0,0 +1,469 @@
+                if new_mp is measurement:
+                    new_mp = new_mp._unflatten(*measurement._flatten())
+                    # if new op isn't queued, need to requeue op.
+                return new_mp
+
+    Now the interpreter can be used to transform functions and jaxpr:
+
+    >>> interpreter = SimplifyInterpreter()
+    >>> def f(x):
+    ...     qml.RX(x, 0)**2
+    ...     qml.adjoint(qml.Z(0))
+    ...     return qml.expval(qml.X(0) + qml.X(0))
+    >>> simplified_f = interpreter(f)
+    >>> print(qml.draw(simplified_f)(0.5)
+    0: ──RX(1.00)──Z─┤  <2.00*X>
+    >>> jaxpr = jax.make_jaxpr(f)(0.5)
+    >>> interpreter.eval(jaxpr.jaxpr, [], 0.5)
+    [expval(2.0 * X(0))]
+
+    **Handling higher order primitives:**
+
+    Two main strategies exist for handling higher order primitives (primitives with jaxpr as metatdata).
+
+    1) Structure preserving. Tracing the execution preserves the higher order primitive.
+    2) Structure flattening. Tracing the execution eliminates the higher order primitive.
+
+    Compilation transforms, like the above ``SimplifyInterpreter``, may prefer to handle higher order primitives
+    via a structure preserving method. After transforming the jaxpr, the `for_loop` still exists. This maintains
+    the compact structure of the jaxpr and reduces the size of the program. This behavior is the default.
+
+    >>> def g(x):
+    ...     @qml.for_loop(3)
+    ...     def loop(i, x):
+    ...         qml.RX(x, 0) ** i
+    ...         return x
+    ...     loop(1.0)
+    ...     return qml.expval(qml.Z(0) + 3*qml.Z(0))
+    >>> jax.make_jaxpr(interpreter(g))(0.5)
+    { lambda ; a:f32[]. let
+        _:f32[] = for_loop[
+        jaxpr_body_fn={ lambda ; b:i32[] c:f32[]. let
+            d:f32[] = convert_element_type[new_dtype=float32 weak_type=True] b
+            e:f32[] = mul c d
+            _:AbstractOperator() = RX[n_wires=1] e 0
+            in (c,) }
+        n_consts=0
+        ] 0 3 1 1.0
+        f:AbstractOperator() = PauliZ[n_wires=1] 0
+        g:AbstractOperator() = SProd[_pauli_rep=4.0 * Z(0)] 4.0 f
+        h:AbstractMeasurement(n_wires=None) = expval_obs g
+    in (h,) }
+
+    Accumulation transforms, like device execution or conversion to tapes, may need to flatten out
+    the higher order primitive to execute it.
+
+    .. code-block:: python
+
+        class AccumulateOps(PlxprInterpreter):
+
+            def __init__(self, ops=None):
+                self.ops = ops
+
+            def setup(self):
+                if self.ops is None:
+                    self.ops = []
+
+            def interpret_operation(self, op):
+                self.ops.append(op)
+
+        @AccumulateOps.register_primitive(qml.capture.primitives.for_loop_prim)
+        def _(self, *invals, jaxpr_body_fn, n_consts):
+            start, stop, step = invals[0], invals[1], invals[2]
+            consts = invals[3 : 3 + n_consts]
+            state = invals[3 + n_consts :]
```

Comment:
Will need updating depending on #6521 as well. This one might be easy to miss as it is a docs example rather than code that will break.

Reply:
updated. thanks for reminding me about the docs.